integration_test/third_party_apps_test.go

```diff
@@ -237,6 +264,158 @@ func runLoggingTestCases(ctx context.Context, logger *logging.DirectoryLogger, v
 	return err
 }
 
+func runMetricsTestCases(ctx context.Context, logger *logging.DirectoryLogger, vm *gce.VM, testCaseBytes []byte) error {
+	var metrics []expectedMetric
+	var err error
+	if err = yaml.UnmarshalStrict(testCaseBytes, &metrics); err != nil {
+		return fmt.Errorf("could not unmarshal contents of expected_metrics.yaml: %v", err)
+	}
+	if err = validateMetrics(metrics); err != nil {
+		return fmt.Errorf("expected_metrics.yaml failed validation: %v", err)
+	}
+	logger.ToMainLog().Printf("Parsed expected_metrics.yaml: %+v", metrics)
+	// Wait for the representative metric first, which is intended to *always*
+	// be sent. If it doesn't exist, we fail fast and skip running the other metrics;
+	// if it does exist, we go on to the other metrics in parallel, by which point they
+	// have gotten a head start and should end up needing fewer API calls before being found.
+	// In both cases we make significantly fewer API calls which helps us stay under quota.
+	for _, metric := range metrics {
+		if !metric.Representative {
+			continue
+		}
+		err = assertMetric(ctx, logger, vm, metric)
+		if gce.IsExhaustedRetriesMetricError(err) {
+			return fmt.Errorf("representative metric %s not found, skipping remaining metrics", metric.Type)
+		}
+		// If err is non-nil here, then the non-representative metric tests later on will
+		// pick it up and report it as part of the multierr.
+		break
+	}
+	// Give some catch-up time to the remaining metrics, which tend to be configured
+	// for a 60-second interval, plus 10 seconds to let the data propagate in the backend.
+	logger.ToMainLog().Println("Found representative metric, sleeping before checking remaining metrics")
+	time.Sleep(70 * time.Second)
+	// Wait for all remaining metrics, skipping the optional ones.
+	// TODO: Improve coverage for optional metrics.
+	//       See https://github.com/GoogleCloudPlatform/ops-agent/issues/486
+	var requiredMetrics []expectedMetric
+	for _, metric := range metrics {
+		if metric.Optional || metric.Representative {
+			logger.ToMainLog().Printf("Skipping optional or representative metric %s", metric.Type)
+			continue
+		}
+		requiredMetrics = append(requiredMetrics, metric)
+	}
+	c := make(chan error, len(requiredMetrics))
+	for _, metric := range requiredMetrics {
+		metric := metric
+		go func() {
+			c <- assertMetric(ctx, logger, vm, metric)
+		}()
+	}
+	for range requiredMetrics {
+		err = multierr.Append(err, <-c)
+	}
+	return err
+}
+
+// validateMetrics checks that all enum fields have valid values and that
+// there is exactly one representative metric in the slice
+func validateMetrics(metrics []expectedMetric) error {
+	var err error
+	// Field validation
+	expectedKinds := []string{"GAUGE", "DELTA", "CUMULATIVE"}
```

Comment:
[Really optional] It should be possible to generate this from the enum, right? E.g., something like:
```
validKindNames := MetricDescriptor_GAUGE.Descriptor().ReservedNames()
for i=1; i < validKindNames.Len(); i++ { // 0 is always UNSPECIFIED
	expectedKinds = append(expectedKinds, validKindNames.Get(i))
}
```
Though the value type enum has extraneous values ([`MONEY`](https://pkg.go.dev/google.golang.org/genproto@v0.0.0-20220317150908-0efb43f6373e/googleapis/api/metric#MetricDescriptor_MONEY)) that would need to be filtered out anyway, so probably not worth it.

Reply:
Will keep for now for simplicity, but I'll keep this in mind -- it might prove useful for other things.